docs/Roadmap.md

```diff
@@ -36,35 +25,75 @@ instance. Care must be taken to keep the standalone version usable.
-  able to make user-named short URLs (e.g., "godbolt.org/u/mattgodbolt/ctad-example"). This goal is a personal pet
-  project of [Matt's](http://github.com/mattgodbolt/).
-- **Multi-file support**. Multiple file compilation units to open the door to seeing LTO and maybe modules. This may
-  include being able to use a `CMake` file to build things.
-- **Modernising the codebase**. Moving the codebase to TypeScript, or something similar that will allow us to worry less
-  about differences between front-end and back-end code (old JavaScript versions), and help us attract more people to
-  the project.
-- **More compilers and libraries**. Plus finishing off the last stragglers of installation.
+### 2023 goals
+
+- **Modernising the codebase**. 2023 has to be the year of all Typescript. We need to continue pushing towards this goal
+  as we are currently blocked using old versions of some libraries due to JS/TS/node.js/ES6 modules conflicts. We're
+  still using an outdated and unsupported module loader and in order to move off it we need to update to a newer
+  loader...that itself doesn't seemingly doesn't support a mixture of typescript and javascript like we currently have.
+  In any case, moving to Typescript has improved our code quality and continues to find bugs, so we should absolutely
+  port to it for this reason alone.
+- **Improved compiler installation directory handling**. This project started in 2022 but left to rot a little, but our
+  existing approach to sharing the 1700+ compilers and libraries between running instances is not scaling well. Behind
+  the scenes we use NFS, and then for performance mount some compilers as read-only squashfs images from NFS. That
+  worked well until we had 1000+ squashfs images to mount, one per compiler. A project to administrate and update a more
+  layered squashfs image containing large numbers of unchanged compilers (think `docker` image layers) was started in
+  2022, but fell behind due to so many other issues, and Matt being pretty rubbish at prioritising things. As nodes are
+  now taking an awful long time to start up, this is becoming a problem.
+- **Improved deployment flow**. The work to deploy Compiler Explorer has increased. At one point I couldn't get a site
+  release out in the length of my train commute to work, which was a tipping point. I took a week off work in 2022 to
+  fix that (amongst other things), but it's still a long and more-tedious-than-I-would-like process. With the recent
+  addition of a GPU node, it's also more tricky than before to keep things updated. I'd like to find a way to get
+  instances to start up without the need to pre-cache the compilation versions (a step we currently require to make
+  startup fast enough). Ideally an update should be able to be directly applied to our staging setup from GitHub, and
+  then promoted to production once things look good.
```

Comment:
I don't think getting rid of discovery step is realistic as long as we have a file-exists-and-version check built into our node app, but open to ideas.

Reply:
I think might also have to spend a bit of research on deploying with switching to a new version group of instances. It doesn't really give much issues how we do it now, so not a priority, but it would be much nicer (and possibly faster).

Reply:
> I don't think getting rid of discovery step is realistic as long as we have a file-exists-and-version check built into our node app, but open to ideas.

Definitely up for thinking more about this. I still think something like:
- a "forever cache" of program output cached on `program, program modification time, args to program, environment vars` - that way we don't need to ever re-run because of CE code changes.
- either: fast-enough squashfs performance to check modification time at startup anyway, OR have squashfs write a quick-to-parse manifest of just the compiler executables on build of an image, so it's built in to the image

Then we "just" parse 1700+ outputs each time. Hopefully that's not too slow, but if it is, we can do it at CI time (provided it didn't require a new "program output")

Or something! argh I promised myself I wouldn't get caught up solutioning :) but here we are...